File: source/javascript/radix_sort.js
function countingSortByDigit(arr, exp) {
    const n = arr.length;
    const output = new Array(n).fill(0);
    const count = new Array(10).fill(0);

    for (let i = 0; i < n; i++) {
        const digit = Math.floor(arr[i] / exp) % 10;
        count[digit]++;
    }
    for (let i = 1; i < 10; i++) count[i] += count[i - 1];
    for (let i = n - 1; i >= 0; i--) {
        const digit = Math.floor(arr[i] / exp) % 10;
        count[digit]--;
        output[count[digit]] = arr[i];
    }
    for (let i = 0; i < n; i++) arr[i] = output[i];
}

function radixSort(arr) {
    const n = arr.length;
    if (n === 0) return arr;

    let maxVal = arr[0];
    for (let i = 1; i < n; i++) {
        if (arr[i] > maxVal) maxVal = arr[i];
    }

    for (let exp = 1; Math.floor(maxVal / exp) > 0; exp *= 10) {
        countingSortByDigit(arr, exp);
    }
    return arr;
}

const arr = [64, 34, 25, 12, 22, 11, 90, 1, 45, 78];
radixSort(arr);
console.log(arr.join(" "));

File: source/rust/radix_sort.rs
fn counting_sort_by_digit(arr: &mut Vec<i32>, exp: i32) {
    let n = arr.len();
    let mut output = vec![0i32; n];
    let mut count = vec![0usize; 10];

    for &x in arr.iter() {
        let digit = ((x / exp) % 10) as usize;
        count[digit] += 1;
    }
    for i in 1..10 {
        count[i] += count[i - 1];
    }
    for i in (0..n).rev() {
        let digit = ((arr[i] / exp) % 10) as usize;
        count[digit] -= 1;
        output[count[digit]] = arr[i];
    }
    arr.copy_from_slice(&output);
}

fn radix_sort(arr: &mut Vec<i32>) {
    if arr.is_empty() { return; }
    let max_val = *arr.iter().max().unwrap();
    let mut exp = 1;
    while max_val / exp > 0 {
        counting_sort_by_digit(arr, exp);
        exp *= 10;
    }
}

fn main() {
    let mut arr = vec![64, 34, 25, 12, 22, 11, 90, 1, 45, 78];
    radix_sort(&mut arr);
    let s: Vec<String> = arr.iter().map(|x| x.to_string()).collect();
    println!("{}", s.join(" "));
}
